File: app/src/main/java/fr/xgouchet/packageexplorer/ui/activities/AppDetailsActivity.java
package fr.xgouchet.packageexplorer.ui.activities;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import butterknife.BindView;
import fr.xgouchet.packageexplorer.R;
import fr.xgouchet.packageexplorer.model.App;
import fr.xgouchet.packageexplorer.ui.fragments.AppDetailFragment;

import static butterknife.ButterKnife.bind;

public class AppDetailsActivity extends AppCompatActivity {

    public static final String EXTRA_PACKAGE_NAME = "package_name";
    private static final String TAG_DETAILS = "details";
    private App app;

    public static void startWithApp(@NonNull Context context, @NonNull App app) {
        Intent intent = buildIntent(context, app);
        context.startActivity(intent);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public static void startWithAppAndTransition(@NonNull Activity activity, @NonNull App app, @NonNull View icon) {
        Intent intent = buildIntent(activity, app);
        icon.setTransitionName("icon");
        ActivityOptionsCompat options = ActivityOptionsCompat.
                makeSceneTransitionAnimation(activity, icon, "icon");
        activity.startActivity(intent, options.toBundle());
    }

    @NonNull
    private static Intent buildIntent(@NonNull Context context, @NonNull App app) {
        Intent intent = new Intent(context, AppDetailsActivity.class);
        intent.putExtra("package_name", app.getPackageName());
        return intent;
    }

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.icon_app)
    ImageView icon;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // handle intent
        handleIntent();
        if (app == null) {
            Toast.makeText(this, "Impossibru !", Toast.LENGTH_SHORT).show();
            finish();
            return;
        }

        // set content
        setContentView(R.layout.activity_app_details);
        bind(this);

        // action bar
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setSubtitle(app.getPackageName());
        }
        icon.setImageDrawable(app.getIcon());
        setTitle(app.getName());


        // details fragment
        Fragment fragment;
        fragment = getSupportFragmentManager().findFragmentByTag(TAG_DETAILS);
        if (fragment == null) {
            fragment = AppDetailFragment.forApp(app.getPackageName());
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.fragment_container, fragment, TAG_DETAILS)
                    .commit();
        }
    }

    private void handleIntent() {
        Intent intent = getIntent();

        String packageName = intent.getStringExtra(EXTRA_PACKAGE_NAME);
        if (!TextUtils.isEmpty(packageName)) {
            app = App.fromPackageName(this, packageName);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            // Respond to the action bar's Up/Home button
            case android.R.id.home:
                supportFinishAfterTransition();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/fr/xgouchet/packageexplorer/appdetails/AppDetailsActivity.kt
package fr.xgouchet.packageexplorer.appdetails

import android.annotation.TargetApi
import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.support.v4.app.ActivityOptionsCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import fr.xgouchet.packageexplorer.R
import fr.xgouchet.packageexplorer.applist.AppViewModel
import fr.xgouchet.packageexplorer.core.utils.Cutelry.knife
import kotlin.properties.Delegates.notNull

/**
 * @author Xavier F. Gouchet
 */
class AppDetailsActivity : AppCompatActivity() {

    companion object {

        val EXTRA_PACKAGE_NAME = "package_name"

        fun startWithAppAndMaybeTransition(activity: Activity,
                                           app: AppViewModel,
                                           transitionView: View?) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
                    && transitionView != null) {
                startWithAppAndTransition(activity, app, transitionView)
            } else {
                startWithApp(activity, app)
            }
        }

        fun startWithApp(activity: Activity, app: AppViewModel) {
            val intent = buildIntent(activity, app)
            activity.startActivity(intent)
        }

        @TargetApi(Build.VERSION_CODES.LOLLIPOP)
        fun startWithAppAndTransition(activity: Activity, app: AppViewModel, transitionView: View) {
            transitionView.transitionName = activity.getString(R.string.transition_icon)
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(activity, transitionView, transitionView.transitionName)
            val intent = buildIntent(activity, app)
            activity.startActivity(intent, options.toBundle())
        }

        fun buildIntent(activity: Activity, app: AppViewModel): Intent {
            val intent = Intent(activity, AppDetailsActivity::class.java)
            intent.putExtra(EXTRA_PACKAGE_NAME, app.packageName)
            return intent
        }
    }

    internal var app: AppViewModel? = null
    internal val icon: ImageView by knife(R.id.icon_app)
    internal val packageName: TextView by knife(R.id.text_package_name)

    internal var appDetailsPresenter: AppDetailsPresenter by notNull()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_app_details)

        readAppFromIntent()
        if (isFinishing) return

        setupHeader()

        setupMVP()
    }

    private fun readAppFromIntent() {
        val intent = intent

        val packageName = intent.getStringExtra(EXTRA_PACKAGE_NAME)
        if (packageName.isNullOrBlank()) {
            finish()
            return
        }

        app = AppViewModel.fromPackageName(this, packageName)
    }

    private fun setupHeader() {
        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        app?.let {
            title = it.title
            toolbar.subtitle = it.packageName
            toolbar.navigationIcon = it.icon
        }

    }

    private fun setupMVP() {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container) as AppDetailsFragment

        appDetailsPresenter = getExistingPresenter(fragment) ?: AppDetailsPresenter(fragment, this, app?.packageName ?: "", app?.isSystemApp == true)
        fragment.presenter = appDetailsPresenter
    }

    private fun getExistingPresenter(fragment: AppDetailsFragment): AppDetailsPresenter? {
        val lastPresenter = lastCustomNonConfigurationInstance as AppDetailsPresenter? ?: return null

        lastPresenter.displayer = fragment
        return lastPresenter
    }

}